src/trace/harTracer.ts

```diff
@@ -0,0 +1,311 @@
+      this._contextTracers.delete(context);
+      await contextTracer.flush();
+    }
+  }
+}
+
+type HarOptions = {
+  path: string;
+  omitContent?: boolean;
+};
+
+class HarContextTracer {
+  private _options: HarOptions;
+  private _browserName: string;
+  private _log: har.Log;
+  private _pageEntries = new Map<Page, har.Page>();
+  private _entries = new Map<network.Request, har.Entry>();
+  private _lastPage = 0;
+  private _barrierPromises = new Map<Promise<void>, Page>();
+
+  constructor(context: BrowserContext, options: HarOptions) {
+    this._browserName = context._browser._options.name;
+    this._options = options;
+    this._log = {
+      version: '1.2',
+      creator: {
+        name: 'Playwright',
+        version: require('../../package.json')['version'],
+      },
+      browser: {
+        name: context._browser._options.name,
+        version: context._browser.version()
+      },
+      pages: [],
+      entries: []
+    };
+    context.on(BrowserContext.Events.Page, page => this._onPage(page));
+  }
+
+  private _onPage(page: Page) {
+    const pageEntry: har.Page = {
+      startedDateTime: new Date(),
+      id: `page_${this._lastPage++}`,
+      title: '',
+      pageTimings: {
+        onContentLoad: -1,
+        onLoad: -1,
+      },
+    };
+    this._pageEntries.set(page, pageEntry);
+    this._log.pages.push(pageEntry);
+    page.on(Page.Events.Request, (request: network.Request) => this._onRequest(page, request));
+    page.on(Page.Events.Response, (response: network.Response) => this._onResponse(page, response));
+
+    page.on(Page.Events.DOMContentLoaded, () => {
+      const promise = page.mainFrame()._evaluateExpression(String(() => {
+        return {
+          title: document.title,
+          domContentLoaded: performance.timing.domContentLoadedEventStart,
+        };
+      }), true, undefined).then(result => {
+        pageEntry.title = result.title;
+        pageEntry.pageTimings.onContentLoad = result.domContentLoaded;
+      }).catch(() => {});
+      this._addBarrier(page, promise);
+    });
+    page.on(Page.Events.Load, () => {
+      const promise = page.mainFrame()._evaluateExpression(String(() => {
+        return {
+          title: document.title,
+          loaded: performance.timing.loadEventStart,
+        };
+      }), true, undefined).then(result => {
+        pageEntry.title = result.title;
+        pageEntry.pageTimings.onLoad = result.loaded;
+      }).catch(() => {});
+      this._addBarrier(page, promise);
+    });
+  }
+
+  private _addBarrier(page: Page, promise: Promise<void>) {
+    const race = Promise.race([
+      new Promise(f => page.on('close', () => {
+        this._barrierPromises.delete(race);
+        f();
+      })),
+      promise
+    ]) as Promise<void>;
+    this._barrierPromises.set(race, page);
+  }
+
+  private _onRequest(page: Page, request: network.Request) {
+    const pageEntry = this._pageEntries.get(page)!;
+    const url = new URL(request.url());
```

Comment:
This scares me a bit because it'll crash node on an invalid URL, and the url is untrusted user content. I guess it would be hard to make a URL that the browser accepts but node does not, but url parsing is not trivial.

Reply:
URL constructor is forgiving...